### DISCUSSION TITLE
Add SnifferChangeStateEvent or similar

### DISCUSSION THREAD

**Discussion by AlepandoCR** [PaperMC/Paper#11251] [Ideas]
### Why?
There 're some events for state changes, for example `PufferFishStateChangeEvent` or in case of poses `EntityPoseChangeEvent` but none of them work for detecting states / poses of the Sniffer, DIGGING is in fact in the enum of Poses detected by `EntityPoseChangeEvent` but its meant for the Warden

### Describe the solution you'd like.

Add a new Event Class to detect the Sniffer state change, and make it so `EntityPoseChangeEvent` can detect the Sniffer poses 

### New Event
```Java
  @EventHandler
    public void snifferStateChange(SnifferChangeStateEvent event){
        plugin.getLogger().info("new state: " + event.getNewState());
        plugin.getLogger().info("old state: " + event.getPreviousState());
        plugin.getLogger().info("sniffer: " + event.getSniffer());
        event.setCancelled(true|false);
    }
```

### Detecting poses
```Java
 @EventHandler
    public void detectSnifferDig(EntityPoseChangeEvent event) {
        Bukkit.getLogger().info("Pose change: " + event.getEntity() + " " + event.getPose());
       if(event.getEntity().getPose().equals(Pose.DIGGING)){ //DIGGING pose is meant for Warden so could add a new "SNIFFER_DIGGING" or share the pose
        Bukkit.getLogger().info("Digging detected by " + event.getEntity());
        }
    }
``` 
This are some examples of functionalities 
    

### Ugly alternative

I don't have experience working with paper pr so I cant offer a real solution to this on paper, but using de API you can do a task checking for Sniffers on a exact state

```Java
 public void lookForDiggingSniffer(){
        new BukkitRunnable() {
            List<World> worlds = Bukkit.getWorlds();
            @Override
            public void run() {
                for(World world : worlds){
                    List<Chunk> chunks = List.of(world.getLoadedChunks());
                    for(Chunk chunk : chunks){
                        for(Entity entity : chunk.getEntities()){
                            if(entity instanceof Sniffer sniffer){
                                if(sniffer.getState().equals(Sniffer.State.DIGGING)){
                                    //Do your thing here
                                }
                            }
                        }
                    }
                }
            }
        }.runTaskTimer(plugin, 0, 1);
    }
```

### Other

Theres a list of new mobs that dont have event listeners, such as the Breeze, Armadillo, Allay, Camel and maybe more, itll be great if you guys could check some of them! :D

**Comment by Malfrador** [comment]
This obviously should be called SnifferSniffEvent.
/s

But yeah, indeed not possible with the API currently. It may make sense to have some sort of catch-all event for all of those mobs though, instead of needing to add individual ones for every one (and every new mob)

**Reply by AlepandoCR** [reply]
idk if sniff event because sniffers have more than one state

```Java
 public static enum State {
        IDLING,
        FEELING_HAPPY,
        SCENTING,
        SNIFFING,
        SEARCHING,
        DIGGING,
        RISING;

        private State() {
        }
    }
```
Nothing wrong with adding individual events for each state but ig it aint that necessary when all states can be checked in one event

Maybe im wrong about that approach so lmk

**Reply by electronicboy** [reply]
They are joking, ofc;

**Reply by AlepandoCR** [reply]
my fault 😔

**Comment by electronicboy** [comment]
I believe that the question over this was if we wanted to try to genericized the concept

**Reply by AlepandoCR** [reply]
Ig the `EntityPoseChangeEvent` is a good approach to getting states / poses from all creatures, but yeah if there was a general `EntityStateChangeEvent` itll be crazy good
